internal/campaigns/volume_workspace.go

```diff
@@ -0,0 +1,199 @@
+package campaigns
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"os"
+
+	"github.com/pkg/errors"
+
+	"github.com/sourcegraph/src-cli/internal/api"
+	"github.com/sourcegraph/src-cli/internal/campaigns/graphql"
+	"github.com/sourcegraph/src-cli/internal/exec"
+)
+
+// dockerVolumeWorkspaceCreator creates dockerWorkspace instances.
+type dockerVolumeWorkspaceCreator struct {
+	client api.Client
+}
+
+var _ WorkspaceCreator = &dockerVolumeWorkspaceCreator{}
+
+func (wc *dockerVolumeWorkspaceCreator) Create(ctx context.Context, repo *graphql.Repository) (Workspace, error) {
+	w := &dockerVolumeWorkspace{}
+	return w, w.init(ctx, wc.client, repo)
+}
+
+func (*dockerVolumeWorkspaceCreator) DockerImages() []string {
+	return []string{dockerWorkspaceImage}
+}
+
+// dockerVolumeWorkspace workspaces are placed on Docker volumes (surprise!),
+// and are therefore transparent to the host filesystem. This has performance
+// advantages if bind mounts are slow, such as on Docker for Mac, but could make
+// debugging harder and is slower when it's time to actually retrieve the diff.
+type dockerVolumeWorkspace struct {
+	volume string
+}
+
+var _ Workspace = &dockerVolumeWorkspace{}
+
+func (w *dockerVolumeWorkspace) init(ctx context.Context, client api.Client, repo *graphql.Repository) error {
```

Comment:
Feel free to ignore: I would've expected the content of this method to be in `dockerVolumeWorkspaceCreator.Create` instead of being a method on this. The `.init` method is more OOP-like, I think, and it comes down to taste here, but just wanted to mention this thought in case you were unsure about code organisation and needed a bit of input.

Reply:
I refactored this as part of the removal of `Prepare()` to be on the creator, and with a more broken down set of methods.